File: typescript/src/escape_html.test.ts
import { describe, it, expect } from 'vitest'
import { escapeHtml } from './escape_html.js'

describe('escapeHtml', () => {
  it('Escape HTML', () => {
    const input = "<script>alert('XSS')</script>"
    const expected = '&#60;script&#62;alert(&#39;XSS&#39;)&#60;/script&#62;'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('All characters', () => {
    const input = "test-&<>'\"-test"
    const expected = 'test-&#38;&#60;&#62;&#39;&#34;-test'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('No escape', () => {
    const input = 'test'
    const expected = 'test'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('Empty string', () => {
    const input = ''
    const expected = ''
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('Escape with special characters', () => {
    const input = "test👌áéíóú-&<>'\"-👌test"
    const expected = 'test👌áéíóú-&#38;&#60;&#62;&#39;&#34;-👌test'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('No escape with special characters', () => {
    const input = 'test👌áéíóú'
    const expected = 'test👌áéíóú'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })

  it('Single special character', () => {
    const input = '👌'
    const expected = '👌'
    const got = escapeHtml(input)
    expect(got).toBe(expected)
  })
})

File: golang/escape_html_test.go
package nodx

import "testing"

func TestEscapeHTML(t *testing.T) {
	t.Run("Escape HTML", func(t *testing.T) {
		input := "<script>alert('XSS')</script>"
		expected := "&#60;script&#62;alert(&#39;XSS&#39;)&#60;/script&#62;"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("All characters", func(t *testing.T) {
		input := "test-&<>'\"-test"
		expected := "test-&#38;&#60;&#62;&#39;&#34;-test"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("No escape", func(t *testing.T) {
		input := "test"
		expected := "test"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("Empty string", func(t *testing.T) {
		input := ""
		expected := ""
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("Escape with special characters", func(t *testing.T) {
		input := "test👌áéíóú-&<>'\"-👌test"
		expected := "test👌áéíóú-&#38;&#60;&#62;&#39;&#34;-👌test"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("No escape with special characters", func(t *testing.T) {
		input := "test👌áéíóú"
		expected := "test👌áéíóú"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})

	t.Run("Single special character", func(t *testing.T) {
		input := "👌"
		expected := "👌"
		got := escapeHTML(input)
		if got != expected {
			t.Errorf("Expected %q, got %q", expected, got)
		}
	})
}
